[Spark] Add Delta Connect Python Client

longvu-db opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://github.com/delta-io/delta/blob/master/CONTRIBUTING.md
  2. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP] Your PR title ...'.
  3. Be sure to keep the PR description updated to reflect all changes.
  4. Please write your PR title to summarize what this PR proposes.
  5. If possible, provide a concise example to reproduce the issue for a faster review.
  6. If applicable, include the corresponding issue number in the PR title and link it in the body.
-->

#### Which Delta project/connector is this regarding?
<!--
Please add the component selected below to the beginning of the pull request title
For example: [Spark] Title of my pull request
-->

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description
This PR adds a basic Python client library for Delta Connect that implements the `DeltaTable.forPath`, `DeltaTable.forName`, `DeltaTable.toDF` and `DeltaTable.alias` methods. It adds checks to the non-Connect versions of `DeltaTable.forPath` and `DeltaTable.forName` to detect the usage of a remote spark session, and to delegate to the Connect versions of these methods.

<!--
- Describe what this PR changes.
- Describe why we need the change.
 
If this PR resolves an issue be sure to include "Resolves #XXX" to correctly link and close the issue upon merge.
-->

## How was this patch tested?
Reused the existing tests for `DeltaTable` in `test_deltatable.py`.
<!--
If tests were added, say they were added here. Please make sure to test the changes thoroughly including negative and positive cases if possible.
If the changes were tested in any way other than unit tests, please clarify how you tested step by step (ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future).
If the changes were not tested, please explain why.
-->

## Does this PR introduce _any_ user-facing changes?
No.
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Delta Lake versions or within the unreleased branches such as master.
If no, write 'No'.
-->

longvu-db committed:
Temp 1

longvu-db committed:
Back to normal

longvu-db committed:
Back to normal

longvu-db committed:
Temp

longvu-db committed:
Back to normal

longvu-db committed:
Enable DV

longvu-db committed:
Before improvement

longvu-db committed:
Broadcasting again

longvu-db committed:
Add back pushdown

longvu-db committed:
Remove pushdown

longvu-db committed:
Add broadcast

longvu-db committed:
Remove broadcast

longvu-db committed:
Add predicate pushdown

longvu-db committed:
Remove pushdown

longvu-db committed:
Add broadcast

longvu-db committed:
Disable DV

longvu-db committed:
Enable this to True

longvu-db committed:
Remove broadcast

longvu-db committed:
Add Enable Pushdown

longvu-db committed:
Add Preprocess DV

longvu-db committed:
Tom's changes

longvu-db committed:
Undo Tom's changes

longvu-db committed:
Add Delta Connect

longvu-db committed:
Fix merge conflicts

longvu-db committed:
Add generated python files

longvu-db committed:
Add yaml file

longvu-db committed:
Update

longvu-db committed:
add lint ignore

longvu-db committed:
Remove unnecessary Delta repo

longvu-db committed:
Rename connect to spark-connect

longvu-db committed:
Update

longvu-db committed:
Update script

longvu-db committed:
Add the necessary mypy ignore stuff

longvu-db committed:
Remove debug

longvu-db committed:
Remove ,

longvu-db committed:
Remove unnecessary

longvu-db committed:
Undo debug

longvu-db committed:
Add Delta Relation

longvu-db committed:
Add Delta Connect Server Library

tlm365 committed:
[INFRA] Improve the java style checks log the errors to sbt console (#3115)

Resolves #3067.

On local machine, intentionally create checkstyle errors in module
`kernelDefaults` (for experimental), then run the `build/sbt compile`
and `build/sbt kernelDefaults/test`.

Signed-off-by: Tai Le Manh <manhtai.lmt@gmail.com>

longvu-db committed:
Try this

longvu-db committed:
SBT

longvu-db committed:
Update

longvu-db committed:
Update

longvu-db committed:
Address comments

johanl-db committed:
[Spark] Test type widening compatibility with other Delta features (#3053)

Additional tests covering type widening and:
- Reading CDF
- Column mapping
- Time travel
- RESTORE
- CLONE

Test only

longvu-db committed:
Update

longvu-db committed:
Add mypy-protobuf

longvu-db committed:
Swap places

longvu-db committed:
Polish build.sbt

longvu-db committed:
Move to Spark Master only

johanl-db committed:
[Spark] Allow type widening for all supported type changes with Spark 4.0 (#3024)

This PR adds shims to ungate the remaining type changes that only work
with Spark 4.0 / master. Spark 4.0 contains the required changes to
Parquet readers to be able to read the data after applying the type
changes.

## Description
Extend the list of supported type changes for type widening to include
changes that can be supported with Spark 4.0:
- (byte, short, int) -> long
- float -> double
- date -> timestampNTZ
- (byte, short, int) -> double
- decimal -> decimal (with increased precision/scale that doesn't cause
precision loss)
- (byte, short, int, long) -> decimal

Shims are added to support these changes when compiling against Spark
4.0/master and to only allow `byte` -> `short` - > `int` when compiling
against Spark 3.5.

## How was this patch tested?
Adding test cases for the new type changes in the existing type widening
test suites. The list of supported / unsupported changes covered in
tests differs between Spark 3.5 and Spark 4.0, shims are also provided
to handle this.

## Does this PR introduce _any_ user-facing changes?
Yes: allow using the listed type changes with type widening, either via
`ALTER TABLE CHANGE COLUMN TYPE` or during schema evolution in MERGE and
INSERT.

longvu-db committed:
Try adding buf

longvu-db committed:
Add python3

longvu-db committed:
Address TD's comments

longvu-db committed:
Update

longvu-db committed:
Try another way to install buf

longvu-db committed:
move directory again

longvu-db committed:
Fix build.sbt

longvu-db committed:
Add back missing files

longvu-db committed:
Fix copyright

longvu-db committed:
add back delta connect proto

longvu-db committed:
Update

longvu-db committed:
Try again

longvu-db committed:
Try again with PATH buf

longvu-db committed:
Remove Scala tests

longvu-db committed:
Add black

longvu-db committed:
Disable other checks

longvu-db committed:
Move back

longvu-db committed:
Remove unnecessary comment

longvu-db committed:
Move again

longvu-db committed:
Enable again

longvu-db committed:
Update

longvu-db committed:
Remove Clone

longvu-db committed:
Update

longvu-db committed:
Add Delta Connect Client Python

longvu-db committed:
Remove unrelated stuffs

longvu-db committed:
Add grpcio

longvu-db committed:
Add Return types

longvu-db committed:
Disable scala tests momentarily

longvu-db committed:
Add Delta Connect GRPC-IO

longvu-db committed:
Add googleapis-common-protos

longvu-db committed:
Add Python installs for Spark Master

longvu-db committed:
Remove GRPC

longvu-db committed:
Add missing depencies

longvu-db committed:
Remove Scala tests temporarily

longvu-db committed:
Add project

longvu-db committed:
Resolve Merge conflicts

longvu-db committed:
Untab

longvu-db committed:
Resolve Andreas' comments

longvu-db committed:
Add Delta Connect python tests

longvu-db committed:
Try

longvu-db committed:
Run for connectServer

longvu-db committed:
Typo

longvu-db committed:
Add publish constraints

scottsand-db committed:
Do not compile, test, or publish Spark Connect common/server projects against Spark 3.5

longvu-db committed:
Polish

longvu-db committed:
Nit

scottsand-db commented on .github/workflows/spark_master_test.yaml:
```diff
@@ -42,9 +42,39 @@ jobs:
         # when changing TEST_PARALLELISM_COUNT make sure to also change it in spark_test.yaml
         run: |
-          TEST_PARALLELISM_COUNT=2 build/sbt -DsparkVersion=master "++ ${{ matrix.scala }}" clean spark/test
+          TEST_PARALLELISM_COUNT=2 IS_SPARK_MASTER=true pipenv run python run-tests.py --group connectServer
```
you can't just remove the `spark/test` eh

scottsand-db replied:
why can't you just add `connectServer/test` ?

longvu-db replied:
@scottsand-db I removed `spark/test` that the testing runs faster, I will add it back once I'm done.

longvu-db replied:
@scottsand-db `connectServer/test` Isn't this for Scala tests? I am trying to run Python tests?

scottsand-db replied:
:dope: my bad

longvu-db committed:
Nit

longvu-db committed:
Add .

longvu-db committed:
Remove dead variable

allisonport-db committed:
get python tests to run

allisonport-db committed:
Initial changes for preview release on Spark 4.0 preview

- Upgrades our version to 4.0.0-SNAPSHOT, and updates Mima settings

- Upgrades Scala to be only Scala 2.13, and makes necessary changes to Kernel code

- Changes the default sparkVersion to be "4.0.0-preview1" and adds the Spark RC repo as a resolver
   - Makes Kernel work with Java 17

- Comments out uniform projects as they require other dependencies

- Comments out flink as it does not support Scala 2.13

- Changes other remaining projects to "skipReleaseSettings" as we decided for now to only release Spark, Kernel, and any storage related projects (due to dependency complexities with Spark and Scala versions)

- Updates our CI jobs for Scala and Java versions

- Updates Spark examples to use the Spark RC

allisonport-db committed:
unidoc fixes

allisonport-db committed:
integration test changes

allisonport-db committed:
Fix sharing and upgrade hadoop version

longvu-db committed:
Merge Allison's changes

longvu-db committed:
Try again

longvu-db committed:
Upgrade pandas

longvu-db committed:
Update PyArrow

longvu-db committed:
Reorganize the imports

longvu-db committed:
Resolve Conflicts

longvu-db committed:
Remove duplicates

longvu-db committed:
Empty commit

longvu-db committed:
Remove random files

longvu-db committed:
Add Back

longvu-db committed:
Space

longvu-db committed:
Change links

longvu-db committed:
Squashed Delta Connect Python Client

longvu-db committed:
Add missing __init__.py file

longvu-db committed:
Merge branch 'delta-connect-python-client' of github.com:longvu-db/delta into delta-connect-python-client

foss-contributor commented:
@longvu-db, thank you for your contribution! This pull request has been inactive for over 18 months and will be closed in one week. To keep it open, please leave a comment on this pull request or contact us in the `#contributors` channel on [Delta Lake Slack](https://go.delta.io/slack). We appreciate your involvement in Delta Lake!
